Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/CancellationRate.hs

```diff
@@ -88,20 +99,72 @@ getCancellationRateData ::
   (MonadFlow m, CacheFlow m r, EsqDBFlow m r) =>
   Id DMOC.MerchantOperatingCity ->
   Id DP.Person ->
-  m CancellationRateData
+  m (Maybe CancellationRateData)
 getCancellationRateData mocId driverId = do
   merchantConfig <- CTC.findByMerchantOpCityId mocId Nothing >>= fromMaybeM (TransporterConfigNotFound mocId.getId)
   let minimumRides = findMinimumRides merchantConfig
-  (assignedCount, cancelledCount, cancellationRate, windowSize) <- do
-    let windowSize = findWindowSize merchantConfig
-    assignedCount <- getAssignedCount windowSize driverId
-    if (isJust merchantConfig.cancellationRateWindow) && (assignedCount > minimumRides)
-      then do
-        cancelledCount <- getCancellationCount windowSize driverId
-        let cancellationRate = (cancelledCount * 100) `div` max 1 assignedCount
-        pure (Just $ fromInteger assignedCount, Just $ fromInteger cancelledCount, Just $ fromInteger cancellationRate, merchantConfig.cancellationRateWindow)
-      else pure (Nothing, Nothing, Nothing, Nothing)
-  pure $ CancellationRateData assignedCount cancelledCount cancellationRate windowSize
+  let windowSize = findWindowSize merchantConfig
+  assignedCount <- getAssignedCount windowSize driverId
+  if (isJust merchantConfig.cancellationRateWindow) && (assignedCount > minimumRides)
+    then do
+      cancelledCount <- getCancellationCount windowSize driverId
+      let cancellationRate = (cancelledCount * 100) `div` max 1 assignedCount
+      pure $
+        Just $
+          CancellationRateData
+            { assignedCount = fromInteger assignedCount,
+              cancelledCount = fromInteger cancelledCount,
+              cancellationRate = fromInteger cancellationRate,
+              windowSize = fromMaybe 7 merchantConfig.cancellationRateWindow
+            }
+    else pure Nothing
   where
     findWindowSize merchantConfig = toInteger $ fromMaybe 7 merchantConfig.cancellationRateWindow
     findMinimumRides merchantConfig = toInteger $ fromMaybe 5 merchantConfig.cancellationRateCalculationThreshold
+
+nudgeOrBlockDriver ::
+  (MonadFlow m, CacheFlow m r, EsqDBFlow m r, CoreMetrics m, HasLocationService m r) =>
+  TransporterConfig ->
+  Id DP.Person ->
+  m ()
+nudgeOrBlockDriver transporterConfig driverId = do
+  case (transporterConfig.cancellationRateWindow, transporterConfig.cancellationRateBasedNudgingAndBlockingConfig) of
+    (Just windowSize, Just CancellationRateBasedNudgingAndBlockingConfig {..}) -> do
+      driver <- QPerson.findById driverId >>= fromMaybeM (PersonNotFound driverId.getId)
```

Comment:
Please pass the driver from very top instead of fetching it again.

Reply:
Sure done